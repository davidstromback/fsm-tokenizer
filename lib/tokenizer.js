import { write } from "./context.js";
import { add, assign, clone, empty, point } from "./point.js";
const matchWhitespace = /\s/;
const matchLineBreak = /\n/;
function defaultTokenFactory(type, value, start, end, offset) {
    return {
        type,
        value,
        start: add(clone(start), offset),
        end: add(clone(end), offset),
    };
}
const defaultOptions = {};
export function tokenizer(schema, createToken = defaultTokenFactory) {
    return function* tokenize(input, options = defaultOptions) {
        var _a;
        const { state = schema.initialState, offset = empty } = options;
        const context = {
            char: "",
            contentEnd: point(-1),
            contentStart: point(),
            createToken,
            location: point(),
            next: undefined,
            offset,
            paddingEnd: point(-1),
            paddingStart: point(),
            state: schema.states[state],
            string: input,
            token: undefined,
        };
        for (let index = 0; index < input.length; index++) {
            context.char = input[index];
            context.token = undefined;
            if (context.next) {
                write(context);
                context.state = context.next;
                context.next = undefined;
            }
            context.state = context.state(context);
            assign(context.paddingEnd, context.location);
            context.location.offset++;
            if (matchLineBreak.test(context.char)) {
                context.location.column = 0;
                context.location.line++;
            }
            else {
                context.location.column++;
            }
            if (matchWhitespace.test(context.char)) {
                if (context.contentEnd.offset < context.paddingStart.offset) {
                    assign(context.contentStart, context.location);
                }
            }
            else {
                assign(context.contentEnd, context.paddingEnd);
            }
            if (context.token) {
                yield context.token;
            }
        }
        write(context);
        if (context.token) {
            yield context.token;
        }
        return {
            state: ((_a = context.next) !== null && _a !== void 0 ? _a : context.state).key,
            offset: add(context.location, context.offset),
        };
    };
}
