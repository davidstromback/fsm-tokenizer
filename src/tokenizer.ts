import type {
  Point,
  Token,
  TokenFactory,
  Context,
  Tokenizer,
  Schema,
} from "./types.js";

import { write } from "./context.js";
import { clone, assign } from "./point.js";

const matchWhitespace = /\s/;
const matchLineBreak = /\n/;

function defaultTokenFactory<T>(
  type: T,
  value: string | undefined,
  start: Point,
  end: Point
): Token<T> {
  return { type, value, start: clone(start), end: clone(end) };
}

export function tokenizer<
  TokenType extends keyof any,
  StateName extends keyof any,
  T = Token<TokenType>
>(
  schema: Schema<TokenType, StateName>,
  createToken?: TokenFactory<T, TokenType>
): Tokenizer<T, StateName>;

export function tokenizer(
  schema: Schema<string, string>,
  createToken = defaultTokenFactory
): Tokenizer<Record<string, any>, string> {
  return function* tokenize(input, initialState = schema.initialState) {
    const context: Context = {
      char: "",
      contentEnd: { line: 1, column: 1, offset: -1 },
      contentStart: { line: 1, column: 1, offset: 0 },
      location: { line: 1, column: 1, offset: 0 },
      next: undefined,
      paddingEnd: { line: 1, column: 1, offset: -1 },
      paddingStart: { line: 1, column: 1, offset: 0 },
      state: schema.states[initialState],
      string: input,
      token: undefined,
      createToken,
    };

    for (let index = 0; index < input.length; index++) {
      context.char = input[index];
      context.token = undefined;

      if (context.next) {
        write(context);
        context.state = context.next;
        context.next = undefined;
      }

      context.state = context.state(context);

      assign(context.paddingEnd, context.location);

      context.location.offset++;

      if (matchLineBreak.test(context.char)) {
        context.location.column = 1;
        context.location.line++;
      } else {
        context.location.column++;
      }

      if (matchWhitespace.test(context.char)) {
        if (context.contentEnd.offset < context.paddingStart.offset) {
          assign(context.contentStart, context.location);
        }
      } else {
        assign(context.contentEnd, context.paddingEnd);
      }

      if (context.token) {
        yield context.token;
      }
    }

    write(context);

    if (context.token) {
      yield context.token;
    }

    return (context.next ?? context.state).key;
  };
}
