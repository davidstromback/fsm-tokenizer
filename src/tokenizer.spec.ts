import { markup } from "./markup.js";
import { tokenizer } from "./tokenizer.js";
import { valid, invalid, split } from "./__tests__/cases.js";
import { tokenizeResult } from "./__tests__/fixtures.js";

describe("tokenizer", () => {
  describe("matches snapshots", () => {
    const tokenize = tokenizer(markup);

    test.each(valid)("%p", (input) => {
      expect(tokenizeResult(tokenize, input)).toMatchSnapshot();
    });
  });

  describe("throws errors matching snapshots", () => {
    const tokenize = tokenizer(markup);

    test.each(invalid)("%p", (input) => {
      expect(() => [...tokenize(input)]).toThrowErrorMatchingSnapshot();
    });
  });

  describe("yields equal results for split strings", () => {
    const tokenize = tokenizer(markup, (type, value) => ({
      type,
      value,
    }));

    test.each(split)("%p", (original, parts) => {
      expect(tokenizeResult(tokenize, ...parts)).toEqual(
        tokenizeResult(tokenize, original)
      );
    });
  });
});
