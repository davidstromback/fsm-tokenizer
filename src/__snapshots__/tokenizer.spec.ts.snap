// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenizer matches snapshots "<bar foo=\\"bar\\"> </bar>" 1`] = `
Object {
  "state": "text",
  "tokens": Array [
    Object {
      "end": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "type": "tagOpener",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "type": "name",
      "value": "bar",
    },
    Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "start": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "type": "name",
      "value": "foo",
    },
    Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "type": "equals",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "start": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "type": "quote",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "start": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "type": "value",
      "value": "bar",
    },
    Object {
      "end": Object {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "start": Object {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "type": "quote",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "start": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "type": "tagCloser",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 17,
        "line": 1,
        "offset": 16,
      },
      "start": Object {
        "column": 17,
        "line": 1,
        "offset": 16,
      },
      "type": "tagOpener",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 18,
        "line": 1,
        "offset": 17,
      },
      "start": Object {
        "column": 18,
        "line": 1,
        "offset": 17,
      },
      "type": "closeMarker",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 21,
        "line": 1,
        "offset": 20,
      },
      "start": Object {
        "column": 19,
        "line": 1,
        "offset": 18,
      },
      "type": "name",
      "value": "bar",
    },
    Object {
      "end": Object {
        "column": 22,
        "line": 1,
        "offset": 21,
      },
      "start": Object {
        "column": 22,
        "line": 1,
        "offset": 21,
      },
      "type": "tagCloser",
      "value": undefined,
    },
  ],
}
`;

exports[`tokenizer matches snapshots "<foo bar />" 1`] = `
Object {
  "state": "text",
  "tokens": Array [
    Object {
      "end": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "type": "tagOpener",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "type": "name",
      "value": "foo",
    },
    Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "start": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "type": "name",
      "value": "bar",
    },
    Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "start": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "type": "closeMarker",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "start": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "type": "tagCloser",
      "value": undefined,
    },
  ],
}
`;

exports[`tokenizer matches snapshots "<foo bar= />" 1`] = `
Object {
  "state": "text",
  "tokens": Array [
    Object {
      "end": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "type": "tagOpener",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "type": "name",
      "value": "foo",
    },
    Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "start": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "type": "name",
      "value": "bar",
    },
    Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "type": "equals",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "start": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "type": "closeMarker",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "type": "tagCloser",
      "value": undefined,
    },
  ],
}
`;

exports[`tokenizer matches snapshots "<foo bar= foobar />" 1`] = `
Object {
  "state": "text",
  "tokens": Array [
    Object {
      "end": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "type": "tagOpener",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "type": "name",
      "value": "foo",
    },
    Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "start": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "type": "name",
      "value": "bar",
    },
    Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "type": "equals",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 16,
        "line": 1,
        "offset": 15,
      },
      "start": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "type": "name",
      "value": "foobar",
    },
    Object {
      "end": Object {
        "column": 18,
        "line": 1,
        "offset": 17,
      },
      "start": Object {
        "column": 18,
        "line": 1,
        "offset": 17,
      },
      "type": "closeMarker",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 19,
        "line": 1,
        "offset": 18,
      },
      "start": Object {
        "column": 19,
        "line": 1,
        "offset": 18,
      },
      "type": "tagCloser",
      "value": undefined,
    },
  ],
}
`;

exports[`tokenizer matches snapshots "<foo bar=\\" \\" />" 1`] = `
Object {
  "state": "text",
  "tokens": Array [
    Object {
      "end": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "type": "tagOpener",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "type": "name",
      "value": "foo",
    },
    Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "start": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "type": "name",
      "value": "bar",
    },
    Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "type": "equals",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "start": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "type": "quote",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "start": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "type": "value",
      "value": " ",
    },
    Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "type": "quote",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "start": Object {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "type": "closeMarker",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "start": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "type": "tagCloser",
      "value": undefined,
    },
  ],
}
`;

exports[`tokenizer matches snapshots "<foo bar=\\"foobar\\" />" 1`] = `
Object {
  "state": "text",
  "tokens": Array [
    Object {
      "end": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "type": "tagOpener",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "type": "name",
      "value": "foo",
    },
    Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "start": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "type": "name",
      "value": "bar",
    },
    Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "type": "equals",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "start": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "type": "quote",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 16,
        "line": 1,
        "offset": 15,
      },
      "start": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "type": "value",
      "value": "foobar",
    },
    Object {
      "end": Object {
        "column": 17,
        "line": 1,
        "offset": 16,
      },
      "start": Object {
        "column": 17,
        "line": 1,
        "offset": 16,
      },
      "type": "quote",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 19,
        "line": 1,
        "offset": 18,
      },
      "start": Object {
        "column": 19,
        "line": 1,
        "offset": 18,
      },
      "type": "closeMarker",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 20,
        "line": 1,
        "offset": 19,
      },
      "start": Object {
        "column": 20,
        "line": 1,
        "offset": 19,
      },
      "type": "tagCloser",
      "value": undefined,
    },
  ],
}
`;

exports[`tokenizer matches snapshots "<foo>text</foo>" 1`] = `
Object {
  "state": "text",
  "tokens": Array [
    Object {
      "end": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "type": "tagOpener",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "type": "name",
      "value": "foo",
    },
    Object {
      "end": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
      "start": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
      "type": "tagCloser",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "start": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "type": "text",
      "value": "text",
    },
    Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "start": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "type": "tagOpener",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "start": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "type": "closeMarker",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "start": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "type": "name",
      "value": "foo",
    },
    Object {
      "end": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "start": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "type": "tagCloser",
      "value": undefined,
    },
  ],
}
`;

exports[`tokenizer matches snapshots "<foobar />" 1`] = `
Object {
  "state": "text",
  "tokens": Array [
    Object {
      "end": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "type": "tagOpener",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "type": "name",
      "value": "foobar",
    },
    Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "type": "closeMarker",
      "value": undefined,
    },
    Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "start": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "type": "tagCloser",
      "value": undefined,
    },
  ],
}
`;

exports[`tokenizer throws errors matching snapshots "<<" 1`] = `"Invalid \\"<\\" at line 1, column 2, expected \\"\\\\s\\", \\"[a-zA-Z]\\", \\"\\\\/\\", \\">\\"."`;

exports[`tokenizer throws errors matching snapshots "<\\"" 1`] = `"Invalid \\"\\"\\" at line 1, column 2, expected \\"\\\\s\\", \\"[a-zA-Z]\\", \\"\\\\/\\", \\">\\"."`;

exports[`tokenizer throws errors matching snapshots "<0>" 1`] = `"Invalid \\"0\\" at line 1, column 2, expected \\"\\\\s\\", \\"[a-zA-Z]\\", \\"\\\\/\\", \\">\\"."`;

exports[`tokenizer throws errors matching snapshots "<foo 0foo />" 1`] = `"Invalid \\"0\\" at line 1, column 6, expected \\"\\\\s\\", \\"[a-zA-Z]\\", \\"\\\\/\\", \\">\\"."`;
